### PR TITLE
More dependency reduction.

### PR DESCRIPTION
**Pull request by samth**
This is now implemented, and is roughly ready for review. Still to-do:
 - [x] run tests
 - [x] run `raco setup`

### CODE REVIEW

**Comment by stamourv** [comment]
LGTM.

FWIW, `syntax/parse/pre` seems to be undocumented.

**Comment by samth** [comment]
@stamourv I checked with @rmculpepper and he said I can rely on it, but yeah, docs would be nice.

File: typed-racket-lib/typed-racket/typecheck/internal-forms.rkt [hunk 1]
```diff
@@ -111,17 +110,18 @@
...
-           (define-syntax-class name
-             #:auto-nested-attributes
-             #:literal-sets ((internal-literals #:at name))
-             (pattern i:internal^ #:with (lit . body) #'i.value))
-           (define pred
-             (syntax-parser
-               [(~var _ name) #t]
-               [_ #f]))) ...))]))
+     (syntax
+      (begin
```
Comment (takikawa) [inline typed-racket-lib/typed-racket/typecheck/internal-forms.rkt]: Can we retain the spacing here to reduce the diff? I think aside from `template` -> `syntax` this is just indentation.

File: typed-racket-lib/typed-racket/typed-racket.rkt [hunk 2]
```diff
@@ -4,7 +4,7 @@
  (for-syntax racket/base racket/lazy-require
              "standard-inits.rkt")
  ;; these need to be available to the generated code
- "typecheck/renamer.rkt"
+ "typecheck/renamer.rkt" syntax/location
```
Comment (takikawa) [inline typed-racket-lib/typed-racket/typed-racket.rkt]: Why is this import necessary? (nothing else in the file seemed to change?)

File: typed-racket-lib/typed-racket/utils/redirect-contract.rkt [hunk 3]
```diff
@@ -1,6 +1,6 @@
 #lang racket/base
 
-(require syntax/modcollapse (for-template racket/base))
+(require syntax/private/modcollapse-noctc (for-template racket/base))
```
Comment (takikawa) [inline typed-racket-lib/typed-racket/utils/redirect-contract.rkt]: How reliable is this require of a private module?

File: typed-racket-lib/typed-racket/typed-racket.rkt (hunk 2 above)
Comment (samth) [inline typed-racket-lib/typed-racket/typed-racket.rkt]: Because we produce generated code (for the blame object for provided identifiers) that uses `syntax/location` -- that's why it's here with the comment. Previously we got away without this because something else depended on `syntax/location`.

File: typed-racket-lib/typed-racket/utils/tc-utils.rkt [hunk 4]
```diff
@@ -6,9 +6,10 @@ don't depend on any other portion of the system
 |#
 
 (require syntax/source-syntax "disappeared-use.rkt"
-         racket/promise racket/string
-         syntax/parse/pre (for-syntax racket/base syntax/parse/pre)
-         (only-in unstable/sequence in-slice))
+         racket/promise racket/string racket/lazy-require
+         syntax/parse/pre (for-syntax racket/base syntax/parse/pre))
+
+(lazy-require [unstable/sequence (in-slice)])
```
Comment (takikawa) [inline typed-racket-lib/typed-racket/utils/tc-utils.rkt]: It shouldn't be too hard to re-write `tc-error/fields` to just avoid this dependency entirely, I think. Though I guess it's not a big deal if it's only required on error?

File: typed-racket-lib/typed-racket/typecheck/internal-forms.rkt (hunk 1 above)
Comment (samth) [inline typed-racket-lib/typed-racket/typecheck/internal-forms.rkt]: Well, really we should fix `template`'s indentation.

File: typed-racket-lib/typed-racket/utils/tc-utils.rkt (hunk 4 above)
Comment (samth) [inline typed-racket-lib/typed-racket/utils/tc-utils.rkt]: The point of this commit is for it not to be `require`d at runtime, so I don't think there's a problem here. I expect that the typechecker itself depends on `unstable/syntax` in a variety of other ways too.

The better solution would be to split `tc-utils` into a module needed only in the type checker and one that gets used by other modules.

File: typed-racket-lib/typed-racket/utils/redirect-contract.rkt (hunk 3 above)
Comment (samth) [inline typed-racket-lib/typed-racket/utils/redirect-contract.rkt]: Currently, `scribble/srcdoc` uses it as well.
